File: golang/eval/fun_def_list.go
package eval

import "github.com/Jumpaku/eyamlate/golang/pb/ast"

func (l *FunDefList) Register(def *ast.FunDef) *FunDefList {
	return &FunDefList{
		Parent: l,
		Def:    def,
	}
}

func (l *FunDefList) Find(ident string) *FunDefList {
	cur := l
	for {
		if cur == nil {
			return nil
		}
		if cur.Def.Def == ident {
			return cur
		}
		cur = cur.Parent
	}
}

File: typescript/src/eval/fun_def_list.ts
import {FunDef, FunDefList, FunDefListSchema} from "./evaluator_pb.js";
import {create} from "@bufbuild/protobuf";


export function empty(): FunDefList {
    return create(FunDefListSchema);
}

export function register(l: FunDefList, def: FunDef): FunDefList {
    return create(FunDefListSchema, {parent: l, def});
}

export function find(l: FunDefList, ident: string): FunDefList | null {
    let cur: FunDefList | null = l;
    while (true) {
        if (cur == null) {
            return null;
        }
        if (cur.def?.def === ident) {
            return cur;
        }
        if (cur.parent == null) {
            return null;
        }
        cur = cur.parent;
    }
}

